backend/src/stwo/prover.rs

```diff
@@ -18,15 +35,102 @@ impl<F: FieldElement> StwoProver<F> {
+        witgen_callback: WitgenCallback<F>,
+    ) -> Result<Vec<u8>, String> {
+        let config = PcsConfig {
+            pow_bits: 16,                          // Any value you want to set for pow_bits
+            fri_config: FriConfig::new(0, 1, 100), // Using different numbers for FriConfig
+        };
+
+        // Precompute twiddles.
+        let twiddles = SimdBackend::precompute_twiddles(
+            CanonicCoset::new(
+                (self.analyzed.degree() as u32) + 1 + config.fri_config.log_blowup_factor,
+            )
+            .circle_domain()
+            .half_coset,
+        );
+
+        // Setup protocol.
+        let prover_channel = &mut Poseidon252Channel::default();
+        let commitment_scheme =
+            &mut CommitmentSchemeProver::<SimdBackend, Poseidon252MerkleChannel>::new(
+                config, &twiddles,
+            );
+
+        let trace = PowdrCircuit::new(self.analyzed.clone())
+            .with_witgen_callback(witgen_callback)
+            .with_witness(witness)
+            .into_stwo_circuit_trace();
+
+        let mut tree_builder = commitment_scheme.tree_builder();
+        tree_builder.extend_evals(trace);
+        tree_builder.commit(prover_channel);
+
+        let component = PowdrComponent::new(
+            &mut TraceLocationAllocator::default(),
+            PowdrEval::new(
+                self.analyzed.clone(),
+                self.analyzed.commitment_count()
+                    + self.analyzed.constant_count()
+                    + self.analyzed.publics_count(),
+            ),
+        );
+
+        //let start = Instant::now();
+        let proof = stwo_prover::core::prover::prove::<SimdBackend, Poseidon252MerkleChannel>(
+            &[&component],
+            prover_channel,
+            commitment_scheme,
+        )
+        .unwrap();
+
+        Ok(bincode::serialize(&proof).unwrap())
+    }
+
+    pub fn verify(&self, proof: &[u8], _instances: &[F]) -> Result<(), String> {
+        let proof: StarkProof<<Poseidon252MerkleChannel as MerkleChannel>::H> =
+            bincode::deserialize(proof).map_err(|e| format!("Failed to deserialize proof: {e}"))?;
+
+        let config = PcsConfig {
+            pow_bits: 16,                          // Any value you want to set for pow_bits
+            fri_config: FriConfig::new(0, 1, 100), // Using different numbers for FriConfig
+        };
+
+        let verifier_channel = &mut Poseidon252Channel::default();
+        let commitment_scheme =
+            &mut CommitmentSchemeVerifier::<Poseidon252MerkleChannel>::new(config);
+
+        //Constraints that are to be proved
+        let component = PowdrComponent::new(
+            &mut TraceLocationAllocator::default(),
+            PowdrEval::new(
+                self.analyzed.clone(),
+                self.analyzed.commitment_count()
+                    + self.analyzed.constant_count()
+                    + self.analyzed.publics_count(),
```

Comment:
cf above

Reply:
prove function of StwoProver is generic for "backend, Merkelchannel, and channel" now.  

in the create() function of StwoProverFactory in mod.rs, StwoProver is initialised as 
```
stwo: Box<StwoProver<F, SimdBackend, Blake2sMerkleChannel, Blake2sChannel>> =
            Box::new(StwoProver::new(pil, fixed)?); 
```
generic type can be specified here.  
Fri parameters are set in prover.rs, store in StwoProver struct so it can be accessed by prove and verify functions. 

I will chek if make a separate struct StwoProverConfig for choosing hash and choosing FRI parameters would be better

Reply:
I see, you implemented the fully generic way already, that's fine I guess!